File: services/alp-db-svc/src/dao/UserDAO.ts
import async from "async";
import * as utils from "../utils/baseUtils";
import * as config from "../utils/config";

class UserDAO {
  private logger = config.getLogger();

  checkIfUserExist = (db: any, user: string) => {
    return new Promise((resolve, reject) => {
      db.executeQuery(
        `Select * from "SYS"."USERS" where user_name = ?`,
        [{ value: user }],
        (err: any, result: any) => {
          if (err) return reject(err);
          if (result && result.length == 1) {
            resolve(true);
          } else {
            resolve(false);
          }
        }
      );
    });
  };

  checkIfRoleExist = (db: any, roleName: string) => {
    return new Promise((resolve, reject) => {
      db.executeQuery(
        `Select * from "SYS"."ROLES" where role_name = ?`,
        [{ value: roleName }],
        (err: any, result: any) => {
          if (err) return reject(err);
          if (result && result.length == 1) {
            resolve(true);
          } else {
            resolve(false);
          }
        }
      );
    });
  };

  createSchemaReadRole = (db: any, schema: string, roleName: string) => {
    return new Promise((resolve, reject) => {
      async.series(
        [
          // function (callback) {
          //   db.setAutoCommitToFalse(callback);
          // },
          function (callback) {
            db.execute(
              `CREATE ROLE "${roleName}" NO GRANT TO CREATOR`,
              [],
              (err: any, result: any) => {
                if (err) return callback(err);
                const response = utils.createMultiLogs(
                  `${roleName} role Created Successfully`,
                  { role: roleName }
                );
                callback(null, response);
              }
            );
          },
          function (callback) {
            db.execute(
              `GRANT SELECT, EXECUTE, CREATE TEMPORARY TABLE ON SCHEMA ${schema} TO ${roleName}`,
              [],
              (err: any, result: any) => {
                if (err) return reject(err);
                const response = utils.createMultiLogs(
                  `Granted Read privileges Successfully`,
                  {
                    role: roleName,
                    schema,
                    privileges: "SELECT, EXECUTE, CREATE TEMPORARY TABLE",
                  }
                );
                callback(null, response);
              }
            );
          },
          // function (callback) {
          //   db.commit(callback);
          // },
        ],
        // optional callback
        (err, results: any) => {
          if (err) {
            this.logger.error(err);
            reject(err);
          } else {
            resolve(results); //success
          }
          // results is now equal to ['one', 'two']
        }
      );
    });
  };

  createUser(db: any, user: string, password: string) {
    return new Promise((resolve, reject) => {
      async.series(
        [
          function (callback) {
            db.execute(
              `CREATE USER ${user} PASSWORD ${password} NO FORCE_FIRST_PASSWORD_CHANGE`,
              [],
              (err: any, result: any) => {
                if (err) return callback(err);
                const response = utils.createMultiLogs(
                  `${user} User Created Successfully`,
                  { user }
                );
                callback(null, response);
              }
            );
          },
        ],
        // optional callback
        (err, results: any) => {
          if (err) {
            this.logger.error(err);
            reject(err);
          } else {
            resolve(results); //success
          }
          // results is now equal to ['one', 'two']
        }
      );
    });
  }

  createRole(db: any, roleName: string) {
    return new Promise((resolve, reject) => {
      async.series(
        [
          function (callback) {
            db.executeQuery(
              `CREATE ROLE "${roleName}"`,
              [],
              (err: any, result: any) => {
                if (err) return callback(err);
                const response = utils.createMultiLogs(
                  `${roleName} role created successfully`,
                  { role: roleName }
                );
                callback(null, response);
              }
            );
          },
          // function (callback) {
          //   db.commit(callback);
          // },
        ],
        // optional callback
        (err, results: any) => {
          if (err) {
            this.logger.error(err);
            reject(err);
          } else {
            resolve(results); //success
          }
          // results is now equal to ['one', 'two']
        }
      );
    });
  }

  assignRole(db: any, user: string, roleName: string) {
    return new Promise((resolve, reject) => {
      async.series(
        [
          function (callback) {
            db.execute(
              `GRANT "${roleName}" TO ${user}`,
              [],
              (err: any, result: any) => {
                if (err) return callback(err);
                const response = utils.createMultiLogs(
                  `${roleName} role granted to ${user} user successfully`,
                  { user, role: roleName }
                );
                callback(null, response);
              }
            );
          },
          // function (callback) {
          //   db.commit(callback);
          // },
        ],
        // optional callback
        (err, results: any) => {
          if (err) {
            this.logger.error(err);
            reject(err);
          } else {
            resolve(results); //success
          }
          // results is now equal to ['one', 'two']
        }
      );
    });
  }

  createAndAssignRole(db: any, user: string, roleName: string) {
    return new Promise((resolve, reject) => {
      async.series(
        [
          // function (callback) {
          //   db.setAutoCommitToFalse(callback);
          // },
          function (callback) {
            db.execute(
              `CREATE ROLE "${roleName}"`,
              [],
              (err: any, result: any) => {
                if (err) return callback(err);
                const response = utils.createMultiLogs(
                  `${roleName} role Created Successfully`,
                  { role: roleName }
                );
                callback(null, response);
              }
            );
          },
          function (callback) {
            db.execute(
              `GRANT "${roleName}" TO ${user} `,
              [],
              (err: any, result: any) => {
                if (err) return callback(err);
                const response = utils.createMultiLogs(
                  `${roleName} Role Granted to ${user} User Successfully`,
                  { user, role: roleName }
                );
                callback(null, response);
              }
            );
          },
          // function (callback) {
          //   db.commit(callback);
          // },
        ],
        // optional callback
        (err, results: any) => {
          if (err) {
            this.logger.error(err);
            reject(err);
          } else {
            resolve(results); //success
          }
          // results is now equal to ['one', 'two']
        }
      );
    });
  }

  grantReadPrivileges = (db: any, schema: string, roleName: string) => {
    return new Promise((resolve, reject) => {
      db.execute(
        `GRANT SELECT, EXECUTE, CREATE TEMPORARY TABLE ON SCHEMA ${schema} TO ${roleName}`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          const response = utils.createMultiLogs(
            `Granted Read privileges Successfully`,
            {
              role: roleName,
              schema,
              privileges: "SELECT, EXECUTE, CREATE TEMPORARY TABLE",
            }
          );
          resolve(response);
        }
      );
    });
  };

  grantWritePrivileges = (db: any, schema: string, roleName: string) => {
    return new Promise((resolve, reject) => {
      db.execute(
        `GRANT DELETE, EXECUTE, INSERT, SELECT, UPDATE, CREATE TEMPORARY TABLE ON SCHEMA ${schema} TO ${roleName}`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          const response = utils.createMultiLogs(
            `Granted Write privileges Successfully`,
            {
              role: roleName,
              schema,
              privileges:
                "DELETE, EXECUTE, INSERT, SELECT, UPDATE, CREATE TEMPORARY TABLE",
            }
          );
          resolve(response);
        }
      );
    });
  };

  grantCohortWritePrivileges = (db: any, schema: string, roleName: string) => {
    return new Promise((resolve, reject) => {
      async.series(
        [
          function (callback) {
            db.execute(
              `GRANT DELETE, INSERT, UPDATE ON ${schema}.cohort TO ${roleName}`,
              [],
              (err: any, result: any) => {
                if (err) {
                  // If cohort table does not exist, dont throw error
                  if (
                    err.code === "42P01" || // PG cohort relation does not exist
                    err.code === 397 // HANA cohort invalid object name
                  ) {
                    // pass
                  } else {
                    return callback(err);
                  }
                }
                const response = utils.createMultiLogs(
                  `Granted COHORT Table Write privileges Successfully`,
                  {
                    role: roleName,
                    schema,
                    table: "COHORT",
                    privileges: "DELETE, INSERT, UPDATE",
                  }
                );
                callback(null, response);
              }
            );
          },
          function (callback) {
            db.execute(
              `GRANT DELETE, INSERT, UPDATE ON ${schema}.cohort_definition TO ${roleName}`,
              [],
              (err: any, result: any) => {
                if (err) {
                  // If cohort_definition table does not exist, dont throw error
                  if (
                    err.code === "42P01" || // PG cohort_definition relation does not exist
                    err.code === 397 // HANA cohort_definition invalid object name
                  ) {
                    // pass
                  } else {
                    return callback(err);
                  }
                }
                const response = utils.createMultiLogs(
                  `Granted COHORT_DEFINITION Table Write privileges Successfully`,
                  {
                    role: roleName,
                    schema,
                    table: "COHORT_DEFINITION",
                    privileges: "DELETE, INSERT, UPDATE",
                  }
                );
                callback(null, response);
              }
            );
          },
        ],
        (err, results: any) => {
          if (err) {
            this.logger.error(err);
            reject(err);
          } else {
            resolve(results);
          }
        }
      );
    });
  };

  // For now this is only a postgres function
  grantAdminPrivileges = (db: any, schema: string, roleName: string) => {
    return new Promise((resolve, reject) => {
      //Access to schema and create objects within schema
      db.execute(
        `GRANT CREATE, USAGE ON SCHEMA "${schema}" to "${roleName}"`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          else resolve(result);
        }
      );

      //Create/Write Access to current Objects
      db.execute(
        `GRANT ALL ON ALL TABLES IN SCHEMA "${schema}" to "${roleName}" WITH GRANT OPTION`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          else resolve(result);
        }
      );
      db.execute(
        `GRANT ALL ON ALL FUNCTIONS IN SCHEMA "${schema}" to "${roleName}" WITH GRANT OPTION`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          else resolve(result);
        }
      );
      db.execute(
        `GRANT ALL ON ALL SEQUENCES IN SCHEMA "${schema}" to "${roleName}" WITH GRANT OPTION`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          else resolve(result);
        }
      );

      //Create/Write Access to Future Objects
      db.execute(
        `ALTER DEFAULT PRIVILEGES IN SCHEMA "${schema}" GRANT ALL ON TABLES TO "${roleName}" WITH GRANT OPTION`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          else resolve(result);
        }
      );
      db.execute(
        `ALTER DEFAULT PRIVILEGES IN SCHEMA "${schema}" GRANT ALL ON SEQUENCES TO "${roleName}" WITH GRANT OPTION`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          else resolve(result);
        }
      );
      db.execute(
        `ALTER DEFAULT PRIVILEGES IN SCHEMA "${schema}" GRANT ALL ON FUNCTIONS TO "${roleName}" WITH GRANT OPTION`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          else resolve(result);
        }
      );

      const response = utils.createMultiLogs(
        `Granted admin privileges successfully`,
        {
          role: roleName,
          schema,
        }
      );
      resolve(response);
    });
  };

  grantAllPrivileges = (db: any, schema: string, roleName: string) => {
    return new Promise((resolve, reject) => {
      db.execute(
        `GRANT CREATE ANY, CREATE OBJECT STRUCTURED PRIVILEGE, DELETE, DROP, EXECUTE, INDEX, INSERT, SELECT, SELECT CDS METADATA, SELECT METADATA, UPDATE ON SCHEMA ${schema} TO ${roleName}`,
        [],
        (err: any, result: any) => {
          if (err) return reject(err);
          const response = utils.createMultiLogs(
            `Granted All privileges Successfully`,
            {
              role: roleName,
              schema,
              privileges:
                "CREATE ANY, CREATE OBJECT STRUCTURED PRIVILEGE, DELETE, DROP, EXECUTE, INDEX, INSERT, SELECT, SELECT CDS METADATA, SELECT METADATA, UPDATE",
            }
          );
          resolve(response);
        }
      );
    });
  };
}

export default UserDAO;

File: services/alp-dataflow-gen/pysrc/dao/UserDao.py
from alpconnection.dbutils import GetDBConnection, get_db_svc_endpoint_dialect
from sqlalchemy import text, MetaData, inspect
from typing import List
from datetime import datetime


class UserDao():
    def __init__(self, database_code: str, schema_name: str, user_type: str):
        self.database_code = database_code
        self.schema_name = schema_name
        self.engine = GetDBConnection(database_code, user_type)
        self.metadata = MetaData(schema=schema_name)
        self.inspector = inspect(self.engine)
        self.dialect = get_db_svc_endpoint_dialect(self.database_code)

    # Formatted string version
    def check_user_exists(self, user: str) -> bool:
        match self.dialect:
            case 'postgres':
                select_stmt = text(
                    f"select * from pg_user where usename = '{user}'")
            case 'hana':
                select_stmt = text(
                    f"select * from sys.users where user_name = '{user}'")
        with self.engine.connect() as connection:
            print(f"Executing check user exists statement..")
            res = connection.execute(select_stmt).fetchall()
            if res == []:
                return False
            else:
                return True

    def check_role_exists(self, role_name: str) -> bool:
        match self.dialect:
            case 'postgres':
                select_stmt = text(
                    f"select * from pg_roles where rolname = '{role_name}'")
            case 'hana':
                select_stmt = text(
                    f"select * from sys.roles where role_name = '{role_name}'")
        with self.engine.connect() as connection:
            print(f"Executing check role exists statement..")
            res = connection.execute(select_stmt).fetchall()
            if res == []:
                return False
            else:
                return True

    def create_read_role(self, role_name: str):
        match self.dialect:
            case 'postgres':
                create_role_stmt = text(f'CREATE ROLE {role_name}')
            case 'hana':
                create_role_stmt = text(
                    f'CREATE ROLE {role_name} NO GRANT TO CREATOR')
        with self.engine.connect() as connection:
            print("Executing create read role statement..")
            create_role_res = connection.execute(
                create_role_stmt)
            connection.commit()
            print(f"{role_name} role Created Successfully")

    def create_user(self, user: str, password: str):
        match self.dialect:
            case 'postgres':
                create_user_stmt = text(
                    f'CREATE USER {user} WITH PASSWORD "{password}"')
            case 'hana':
                create_user_stmt = text(
                    f'CREATE USER {user} PASSWORD "{password}" NO FORCE_FIRST_PASSWORD_CHANGE')
        with self.engine.connect() as connection:
            print("Executing create user statement..")
            create_user_res = connection.execute(
                create_user_stmt)
            connection.commit()
            print(f"{user} User Created Successfully")

    def create_and_assign_role(self, user: str, role_name: str):
        with self.engine.connect() as connection:
            create_role_stmt = text(f"CREATE ROLE {role_name}")
            print("Executing create role statement..")
            create_role_res = connection.execute(
                create_role_stmt)
            print(f"{role_name} role Created Successfully")

            grant_role_stmt = text(f"GRANT {role_name} TO {user}")
            print("Executing grant role to user statement..")
            grant_role_res = connection.execute(
                grant_role_stmt, {"x": role_name, "y": user})
            connection.commit()
            print(f" {role_name} Role Granted to {user} User Successfully")

    def grant_read_privileges(self, role_name: str):
        match self.dialect:
            case 'postgres':
                grant_read_stmt = text(f"""
                    GRANT USAGE ON SCHEMA {self.schema_name} TO {role_name};
                    GRANT SELECT ON ALL TABLES IN SCHEMA {self.schema_name} TO {role_name};
                    GRANT USAGE, SELECT ON ALL SEQUENCES IN SCHEMA {self.schema_name} TO {role_name};
                    GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA {self.schema_name} TO {role_name};
                    GRANT EXECUTE ON ALL PROCEDURES IN SCHEMA {self.schema_name} TO {role_name};
                    ALTER DEFAULT PRIVILEGES IN SCHEMA {self.schema_name} GRANT SELECT ON TABLES TO {role_name};
                    ALTER DEFAULT PRIVILEGES IN SCHEMA {self.schema_name} GRANT USAGE, SELECT ON SEQUENCES TO {role_name};
                    ALTER DEFAULT PRIVILEGES IN SCHEMA {self.schema_name} GRANT EXECUTE ON FUNCTIONS TO {role_name};""")
            case 'hana':
                grant_read_stmt = text(
                    f"GRANT SELECT, EXECUTE, CREATE TEMPORARY TABLE ON SCHEMA {self.schema_name} to {role_name}")
        with self.engine.connect() as connection:

            print("Executing grant read privilege statement..")
            grant_read_res = connection.execute(
                grant_read_stmt)
            connection.commit()
            print(f"Granted Read privileges Successfully")

    def grant_cohort_write_privileges(self, role_name: str):
        with self.engine.connect() as connection:
            grant_cohort_write_stmt = text(
                f"GRANT DELETE, INSERT, UPDATE ON {self.schema_name}.cohort TO {role_name}")
            grant_cohort_def_write_stmt = text(
                f"GRANT DELETE, INSERT, UPDATE ON {self.schema_name}.cohort_definition TO {role_name}")
            print("Executing grant cohort write privilege statement..")
            try:
                grant_cohort_write_res = connection.execute(
                    grant_cohort_write_stmt)
                grant_cohort_def_write_res = connection.execute(
                    grant_cohort_def_write_stmt)
                connection.commit()
            except Exception as e:
                print(f"e is {e}")
            else:
                print(
                    f"Granted cohort and cohort definition Write privileges Successfully")

    # Parameterized query
    '''
    def check_user_exists(self, user: str) -> bool:
        match self.dialect:
            case 'postgres':
                select_stmt = text(
                    "select * from pg_user where usename = :x")
            case 'hana':
                select_stmt = text(
                    "select * from sys.users where user_name = :x")
        with self.engine.connect() as connection:
            print(f"Executing check user exists statement..")
            res = connection.execute(select_stmt, {"x": user}).fetchall()
            if res == []:
                return False
            else:
                return True

    def check_role_exists(self, role_name: str) -> bool:
        match self.dialect:
            case 'postgres':
                select_stmt = text(
                    "select * from pg_roles where rolname = :x")
            case 'hana':
                select_stmt = text(
                    "select * from sys.roles where role_name = :x")
        with self.engine.connect() as connection:
            print(f"Executing check role exists statement..")
            res = connection.execute(select_stmt, {"x": role_name}).fetchall()
            if res == []:
                return False
            else:
                return True

    def create_read_role(self, role_name: str):
        match self.dialect:
            case 'postgres':
                create_role_stmt = text('CREATE ROLE :x')
            case 'hana':
                create_role_stmt = text('CREATE ROLE :x NO GRANT TO CREATOR')
        with self.engine.connect() as connection:
            print("Executing create read role statement..")
            create_role_res = connection.execute(
                create_role_stmt, {"x": role_name})
            connection.commit()
            print(f"{role_name} role Created Successfully")

    def create_user(self, user: str, password: str):
        with self.engine.connect() as connection:
            create_user_stmt = text(
                'CREATE USER :x PASSWORD ":y" NO FORCE_FIRST_PASSWORD_CHANGE')
            print("Executing create user statement..")
            create_user_res = connection.execute(
                create_user_stmt, {"x": user, "y": password})
            connection.commit()
            print(f"{user} User Created Successfully")

    def create_and_assign_role(self, user: str, role_name: str):
        with self.engine.connect() as connection:
            create_role_stmt = text("CREATE ROLE :x")
            print("Executing create role statement..")
            create_role_res = connection.execute(
                create_role_stmt, {"x": role_name})
            print(f"{role_name} role Created Successfully")
            grant_role_stmt = text("GRANT :x TO :y")
            print("Executing grant role to user statement..")
            grant_role_res = connection.execute(
                grant_role_stmt, {"x": role_name, "y": user})
            connection.commit()
            print(f" {role_name} Role Granted to {user} User Successfully")

    def grant_read_privileges(self, role_name: str):
        match self.dialect:
            case 'postgres':
                grant_read_stmt = text("""
                    GRANT USAGE ON SCHEMA :x TO :y;
                    GRANT SELECT ON ALL TABLES IN SCHEMA :x TO :y;
                    GRANT USAGE, SELECT ON ALL SEQUENCES IN SCHEMA :x TO :y;
                    GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA :x TO :y;
                    GRANT EXECUTE ON ALL PROCEDURES IN SCHEMA :x TO :y;
                    ALTER DEFAULT PRIVILEGES IN SCHEMA :x GRANT SELECT ON TABLES TO :y;
                    ALTER DEFAULT PRIVILEGES IN SCHEMA :x GRANT USAGE, SELECT ON SEQUENCES TO :y;
                    ALTER DEFAULT PRIVILEGES IN SCHEMA :x GRANT EXECUTE ON FUNCTIONS TO :y;""")
            case 'hana':
                grant_read_stmt = text(
                    "GRANT SELECT, EXECUTE, CREATE TEMPORARY TABLE ON SCHEMA :x to :y")
        with self.engine.connect() as connection:

            print("Executing grant read privilege statement..")
            grant_read_res = connection.execute(
                grant_read_stmt, {"x": text(self.schema_name), "y": text(role_name)})
            connection.commit()
            print(f"Granted Read privileges Successfully")

    def grant_cohort_write_privileges(self, role_name: str):
        with self.engine.connect() as connection:
            grant_cohort_write_stmt = text(
                "GRANT DELETE, INSERT, UPDATE ON :x.cohort TO :y")
            grant_cohort_def_write_stmt = text(
                "GRANT DELETE, INSERT, UPDATE ON :x.cohort_definition TO :y")
            print("Executing grant cohort write privilege statement..")
            try:
                grant_cohort_write_res = connection.execute(
                    grant_cohort_write_stmt, {"x": self.schema_name, "y": role_name})
                grant_cohort_def_write_res = connection.execute(
                    grant_cohort_def_write_stmt, {"x": self.schema_name, "y": role_name})
                connection.commit()
            except Exception as e:
                print(f"e is {e}")
            else:
                print(
                    f"Granted cohort and cohort definition Write privileges Successfully")

    '''

    def grant_write_privileges(self):
        # not used
        pass

    def grant_admin_privileges(self):
        # used for create staging area
        pass

    def grant_all_privileges(self):
        # not used
        pass

    def create_role(self, role_name: str):
        # used for create staging area
        pass

    def assign_role(self):
        # used for create staging area
        pass
